### PR TITLE
zmq test: fix sync-up by matching notification to generated block

### LINKED ISSUE #21216: Failure in test/functional/interface_zmq.py
I don't think this can be my bug, as I pushed a comment only change to my PR and then got a test failure:

https://cirrus-ci.com/task/5869686954917888?command=ci#L5219

```
 test  2021-02-16T19:08:03.550000Z TestFramework (ERROR): Assertion failed 
                                   Traceback (most recent call last):
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 127, in main
                                       self.run_test()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_zmq.py", line 82, in run_test
                                       self.test_sequence()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_zmq.py", line 277, in test_sequence
                                       assert_equal((self.nodes[0].getbestblockhash(), "C", None), seq.receive_sequence())
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/util.py", line 50, in assert_equal
                                       raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
                                   AssertionError: not(('5d2daef0df4c3f8af7d33f350253d3c5b3eab0154803f0f9dff228512ce6ced3', 'C', None) == ('3bd98765bf07cc72a921a5c6fad95b13dac7c80100f0b295f29681e0e3c6e856', 'C', None))
```

The PR in question needs a rebase anyway now, so I'll try that and see if this goes away.

### PR DESCRIPTION
**Pull request by theStack**
This is a follow-up PR for #21008, fixes #21216.

In the course of investigating the problem with jnewbery (analyzing the Cirrus log https://cirrus-ci.com/task/4660108304056320), it turned out that the "sync up" procedure of repeatedly generating a block and waiting for a notification with timeout is too brittle in its current form, as the following scenario could happen:

- generate block A
- receive notification, timeout happens => repeat procedure
- generate block B
- node publishes block A notification
- receive notification, we receive the one caused by block A (!!!) => sync-up procedure is completed
- node publishes block B notification
- the actual test starts
- on the first notification reception, the one caused by block B is received, rather than the one actually caused by test code => assertion failure

This change in the PR ensures that after each test block generation, we wait for the notification that is actually caused by that block and ignore others from possibly earlier blocks. The matching is kind of ugly, it assumes that one out of four components in the block is contained in the notification: the block hash, the tx id, the raw block data or the raw transaction data. (Unfortunately we have to support all publisher topics.)

I'm aware that this is quite a lot of code now only for establishing a robust test setup. OTOH I wouldn't know of a better method right now, suggestions are very welcome.

Note for potential reviewers: for both reproducing the issue on master branch and verifying on PR branch, one can simply generate two blocks in the sync-up procedure rather than one.

### CODE REVIEW

File: test/functional/interface_zmq.py [hunk 1]
```diff
@@ -105,17 +125,18 @@ def setup_zmq_test(self, services, *, recv_timeout=60, sync_blocks=True):
...
             sub.socket.set(zmq.RCVTIMEO, 1000)
         while True:
-            self.nodes[0].generate(1)
+            test_block = ZMQTestSetupBlock(self.nodes[0])
             recv_failed = False
             for sub in subscribers:
                 try:
-                    sub.receive()
+                    while not test_block.caused_notification(sub.receive().hex()):
+                        self.log.debug("Ignoring sync-up notification for previoulsy generated block.")
```
Comment (jonatack) [inline test/functional/interface_zmq.py:139]:
```suggestion
                        self.log.debug("Ignoring sync-up notification for previously generated block.")
```

**Review by jonatack — COMMENTED** [review]
Approach ACK, thanks for working on this.

<details><summary>some suggestions</summary><p>

```diff
-# Helper class for setting up a ZMQ test via the "sync up" procedure.
-# Generates a block on the specified node on instantiation and provides a method to check
-# whether a ZMQ notification matches, i.e. the event was caused by this generated block.
-# Assumes that a notification either contains the generated block's hash, it's
-# (coinbase) transaction id, the raw block or raw transaction data.
 class ZMQTestSetupBlock:
+    """Helper class for setting up a ZMQ test via the "sync up" procedure.
+    Generates a block on the specified node on instantiation and provides a
+    method to check whether a ZMQ notification matches, i.e. the event was
+    caused by this generated block. Assumes that a notification either contains
+    the generated block's hash, it's (coinbase) transaction id, the raw block or
+    raw transaction data.
+    """
+
     def __init__(self, node):
         self.block_hash = node.generate(1)[0]
         coinbase = node.getblock(self.block_hash, 2)['tx'][0]
@@ -76,10 +79,12 @@ class ZMQTestSetupBlock:
         self.raw_block = node.getblock(self.block_hash, 0)
 
     def caused_notification(self, notification):
-        return self.block_hash in notification or \
-               self.tx_hash in notification or \
-               self.raw_block in notification or \
-               self.raw_tx in notification
+        return (
+            self.block_hash in notification
+            or self.tx_hash in notification
+            or self.raw_block in notification
+            or self.raw_tx in notification
+        )
 
 
 class ZMQTest (BitcoinTestFramework):
@@ -136,7 +141,7 @@ class ZMQTest (BitcoinTestFramework):
             for sub in subscribers:
                 try:
                     while not test_block.caused_notification(sub.receive().hex()):
-                        self.log.debug("Ignoring sync-up notification for previoulsy generated block.")
+                        self.log.debug("Ignoring sync-up notification for previously generated block.")
                 except zmq.error.Again:
                     self.log.debug("Didn't receive sync-up notification, trying again.")
                     recv_failed = True
@@ -361,7 +366,7 @@ class ZMQTest (BitcoinTestFramework):
             block_count = self.nodes[0].getblockcount()
             best_hash = self.nodes[0].getbestblockhash()
             self.nodes[0].invalidateblock(best_hash)
-            sleep(2) # Bit of room to make sure transaction things happened
+            sleep(2)  # Bit of room to make sure transaction things happened
 
             # Make sure getrawmempool mempool_sequence results aren't "queued" but immediately reflective
             # of the time they were gathered.
@@ -410,8 +415,8 @@ class ZMQTest (BitcoinTestFramework):
             assert_equal(label, "A")
             # More transactions to be simply mined
             for i in range(len(more_tx)):
-                    assert_equal((more_tx[i], "A", mempool_seq), seq.receive_sequence())
-                    mempool_seq += 1
+                assert_equal((more_tx[i], "A", mempool_seq), seq.receive_sequence())
+                mempool_seq += 1
             # Bumped by rbf
             assert_equal((orig_txid, "R", mempool_seq), seq.receive_sequence())
             mempool_seq += 1
@@ -426,7 +431,7 @@ class ZMQTest (BitcoinTestFramework):
             assert_equal((orig_txid_2, "A", mempool_seq), seq.receive_sequence())
             mempool_seq += 1
             self.nodes[0].generatetoaddress(1, ADDRESS_BCRT1_UNSPENDABLE)
-            self.sync_all() # want to make sure we didn't break "consensus" for other tests
+            self.sync_all()  # want to make sure we didn't break "consensus" for other tests
 
     def test_mempool_sync(self):
```
</p></details>

**Comment by theStack** [comment]
Force-pushed with all changes suggested by @jonatack (thanks for catching the typo and the style fixups, looks much better now! 🚀 ).

**Comment by MarcoFalke** [comment]
Concept ACK 8a8c6383f6f9da10b931f00ca1220408fede8f35

Thanks!

**Pull request merged**